File: 2term/jvmWorld/2lab_kotlin_btree/src/main/kotlin/TreeProvider.kt

class TreeProvider {
    private var tree: BinaryTree<CustomTypeTrait>? = null
    private val deserializers: Map<String, CustomTypeTrait> = mapOf(
        GeoCoordinate(0.0, 0.0).typeName to GeoCoordinate(0.0, 0.0),
        Point2D(0.0, 0.0).typeName to Point2D(0.0, 0.0),
    )
    private var currentType: String = ""

    fun getAvailableTypes(): List<String> {
        return deserializers.keys.toList()
    }

    fun getObjectStringRepresentationExample(): String {
        return deserializers[currentType]?.typeExampleRepresentation
            ?: "Unable to get type representation"
    }

    fun insertRandomValue() {
        val deserializer = deserializers[currentType]
            ?: throw IllegalStateException("Could not find requested type")
        tree?.insert(deserializer.createRandomInstance())
            ?: throw IllegalStateException("Tree not initialized")
    }

    fun insertValue(typeName: String, serializedData: String) {
        val deserializer = deserializers[typeName]
            ?: throw IllegalArgumentException("Unsupported type: $typeName")
        val value = deserializer.deserializeFromString(serializedData)
            ?: throw IllegalArgumentException("Could not deserialize data: $serializedData")
        tree?.insert(value)
            ?: throw IllegalStateException("Tree not initialized")
    }

    fun initializeWithEmptyTree(typeName: String) {
        if (deserializers.containsKey(typeName)) {
            tree = BinaryTree<CustomTypeTrait>()
            currentType = typeName
        } else {
            throw IllegalArgumentException("Unsupported type: $typeName")
        }
    }

    fun getTreeStringRepresentation(): String? {

        return tree?.serialize()
    }

//    fun setTreeFromStringRepresentation(s: String): String {
//        val parsedType = Json.plain.parseJson(s).jsonObject["type"]!!.content
//        val deserializer = deserializers[parsedType]
//            ?: throw IllegalArgumentException("Unsupported type: $parsedType")
//        tree = BinaryTree.fromJsonString(s, deserializer)
//        currentType = parsedType
//        return parsedType
//    }

    fun getTree(): BinaryTree<CustomTypeTrait>? {
        return tree
    }
}

File: 2term/jvmWorld/1lab_scala_binaryTree/src/main/scala/TreeProvider.scala

class TreeProvider(var tree: Option[BinaryTree[CustomTypeTrait]] = None) {
  private val deserializers: Map[String, CustomTypeTrait] = Map(
    "GeoCoordinate" -> new GeoCoordinate(0, 0),
    "Point2D" -> new Point2D(0, 0)
  )
  private var currentType: String = ""

  def getAvailableTypes(): List[String] = {
    deserializers.keys.toList
  }

  def getObjectStringRepresentationExample(): String = {
    deserializers.get(currentType) match {
      case Some(deserializer) => deserializer.typeExampleRepresentation
      case None => "Unable to get type representation"
    }
  }

  def insertRandomValue(): Unit = {
    deserializers.get(currentType) match {
      case Some(deserializer) =>
        tree match {
          case Some(binaryTree) => binaryTree.insert(deserializer.createRandomInstance())
          case None => throw new IllegalStateException("Tree not initialized")
        }
      case None => throw new IllegalStateException("Could not find requested type")
    }
  }

  def insertValue(typeName: String, serializedData: String): Unit = {
    deserializers.get(typeName) match {
      case Some(deserializer) =>
        deserializer.deserializeFromString(serializedData) match {
          case Some(value) =>
            tree match {
              case Some(binaryTree) => binaryTree.insert(value.asInstanceOf[CustomTypeTrait])
              case None => throw new IllegalStateException("Tree not initialized")
            }
          case None => throw new IllegalArgumentException(s"Could not deserialize data: $serializedData")
        }
      case None => throw new IllegalArgumentException(s"Unsupported type: $typeName")
    }
  }

  def initializeWithEmptyTree(typeName: String): Unit = {
    deserializers.get(typeName) match {
      case Some(_) =>
        tree = Some(new BinaryTree[CustomTypeTrait])
      case None => throw new IllegalArgumentException(s"Unsupported type: $typeName")
    }

    currentType = typeName
  }


  def getTree(): Option[BinaryTree[CustomTypeTrait]] = tree
}